File: app/src/main/java/com/example/dices2/Game.java
package com.example.dices2;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static com.example.dices2.Consts.*;

public class Game {
    private static Game instance;
    private MainActivity mainActivity;
    private ListActivity listActivity;
    private List<Player> players;
    private Cell currentCell;
    private boolean isCountTotalEveryMove = false;
    private boolean isCountPlaceEveryMove = false;
    private int totalNumberOfMoves;
    // TODO добавить другие виды правил игры
    private int totalNumberOfRows = 16;
    private int idCurrentPlayer = 0;
    private boolean isGameStarted;
    private Player currentPlayer;
    //private Map<Player, Integer> playersPlaces;

    private Game() {
        players = new ArrayList<>();
        savedCells = new HashSet<>();
    }

    public static Game getInstance() {
        if (instance == null) {
            instance = new Game();
        }
        return instance;
    }

    // Ход игры

    private void startGame() {
        getTotalNumberOfMoves();
        mainActivity.fillMainTable(players, currentPlayer);
        idCurrentPlayer = 0;
        highlightNextPlayer();
        clearPlayersValues();
        isGameStarted = true;
        currentPlayer = players.get(idCurrentPlayer);
    }

    private void clearPlayersValues() {
        for(Player player : players) {
            player.clearAllValues();
        }
    }

    private void highlightNextPlayer() {
        int playersAmount = players.size();
        idCurrentPlayer = (idCurrentPlayer > playersAmount - 1) ? 0 : idCurrentPlayer;
        int idHighlightOff = (idCurrentPlayer == 0) ? playersAmount - 1 : idCurrentPlayer - 1;
        int idHighlightOn = (idCurrentPlayer > playersAmount - 1) ? 0 : idCurrentPlayer;
        mainActivity.switchOffPlayersMove(players.get(idHighlightOff));
        currentPlayer = players.get(idHighlightOn);
        mainActivity.highlightPlayersMove(players.get(idHighlightOn));
        idCurrentPlayer++;
    }

    private void makeMove() {
        totalNumberOfMoves--;
        highlightNextPlayer();
        mainActivity.setTextViewHighlight(currentCell, true);
    }

    private void makeEdit(Cell cell) {
        if (currentCell.isEmpty()) {
            mainActivity.setTextViewHighlight(currentCell, false);
        } else {
            mainActivity.setTextViewHighlight(currentCell, true);
        }
        if (cell.getOwner() == currentPlayer) {
            mainActivity.highlightPlayersMove(currentPlayer);
        }
//        if (isEditInSchool(cell)) {
//            mainActivity.disableFullSquarePokerTextViews(currentPlayer);
//        }
    }

    public void okButtonOnKeyboardActivityClicked(String value, boolean isEdit) {
        Player player = currentCell.getOwner();
        RowName rowName = currentCell.getRow();
        player.setValue(rowName, value);
        currentCell.setValue(value);
        refreshCellsList();
        //setTextToCurrentTextView(value);
        refreshFullSquarePokerTextViews(player);
        if (isEdit) {
            makeEdit(currentCell);
        } else {
            makeMove();
        }
        showSchool(player);
        if (isCountTotalEveryMove) {
            showTotal(player);
        }
        if (isCountPlaceEveryMove) {
            showAllPlaces();
        }
        if (isGameOver()) {
            showAllTotals();
            isGameStarted = false;
            showAllPlaces();
        }
    }

    private void refreshFullSquarePokerTextViews(Player player) {
        if (player.isThreeClassesInSchoolAreFinished() && !player.isSecondNonSchoolPartAdded()) {
            mainActivity.enableFullSquarePokerTextViews(player);
        }
        if (!player.isThreeClassesInSchoolAreFinished() && player.isSecondNonSchoolPartAdded()) {
            mainActivity.disableFullSquarePokerTextViews(player);
        }
    }

    public void setMainActivity(MainActivity mainActivity) {
        this.mainActivity = mainActivity;
        mainActivity.fillMainTable(players, currentPlayer);
    }

    public void setCurrentCell(Cell cell) {
        this.currentCell = cell;
    }

//    public void setTextToCurrentTextView(String value) {
//        currentCell.setValue(value);
//    }

    private void showSchool(Player player) {
        mainActivity.showSchool(player);
    }

    private void showTotal(Player player) {
        mainActivity.showTotal(player);
    }

    private void showAllTotals() {
        for (Player player : players) {
            showTotal(player);
        }
    }

    private void showPlayerPlace(Player player) {
        mainActivity.showPlayerPlace(player);
    }

    private void showAllPlaces() {
        calcPlaces();
        for (Player player : players) {
            showPlayerPlace(player);
        }
    }

    // Работа с ListActivity

    public void setListActivity(ListActivity listActivity) {
        this.listActivity = listActivity;
        fillListView();
    }

    private void fillListView() {
        listActivity.fillListView(players);
    }

    public void addPlayer(String nameOfPlayer) {
        players.add(new Player(nameOfPlayer));
        listActivity.refreshListView(players);
    }

    public void removePlayer(int position) {
        players.remove(position);
        listActivity.refreshListView(players);
    }

    public void clearListOfPlayers() {
        players.clear();
        listActivity.refreshListView(players);
    }

    public String getNameOfPlayerByPosition(int position) {
        return players.get(position).getName();
    }

    public void startGameButtonClicked() {
        startGame();
    }

    // Прочее

    private boolean isGameOver() {
        return totalNumberOfMoves <= 0;
    }

    private int getTotalNumberOfMoves() {
        return totalNumberOfMoves = players.size() * totalNumberOfRows;
    }

    public void onEditMenuClicked() {
        Cell cell = currentCell;
        RowName rowName = cell.getRow();
        mainActivity.startKeyboardActivityForEdit(currentCell.getText().toString(), rowName);
    }

    public void setCountTotalEveryMove(boolean countTotalEveryMove) {
        isCountTotalEveryMove = countTotalEveryMove;
    }

    public void setCountPlaceEveryMove(boolean countPlaceEveryMove) {
        isCountPlaceEveryMove = countPlaceEveryMove;
    }

    public boolean isCountTotalEveryMove() {
        return isCountTotalEveryMove;
    }

    public boolean isGameStarted() {
        return isGameStarted;
    }

    public boolean isDuplicatePlayer(String name) {
        for (Player player : players) {
            if (player.getName().equals(name)) {
                return true;
            }
        }
        return false;
    }

    private void calcPlaces() {
        //playersPlaces = new LinkedHashMap<>();
        List<Player> playersSortedByTotalValue = new ArrayList<>(players);
        Collections.sort(playersSortedByTotalValue, Collections.<Player>reverseOrder());
        Player currentPlayer = playersSortedByTotalValue.get(0);
        int place = 1;
        int currentValue = currentPlayer.getTotal();
        currentPlayer.setCurPlace(place);
        //playersPlaces.put(currentPlayer, place);
        for (int i = 1; i < playersSortedByTotalValue.size(); i++) {
            Player nextPlayer = playersSortedByTotalValue.get(i);
            if (currentValue == nextPlayer.getTotal()) {
                //playersPlaces.put(nextPlayer, place);
                nextPlayer.setCurPlace(place);
            } else {
                place++;
                currentValue = nextPlayer.getTotal();
                //playersPlaces.put(nextPlayer, place);
                nextPlayer.setCurPlace(place);
            }
        }
    }

    public boolean isCountPlaceEveryMove() {
        return isCountPlaceEveryMove;
    }

    private boolean isEditInSchool(Cell cell) {
        for (RowName rowName : NAMES_OF_SCHOOL_ROWS) {
            if (cell.getRow() == rowName) {
                return true;
            }
        }
        return false;
    }

    private void refreshCellsList() {
        savedCells.add(currentCell);
    }

    public boolean getIsCountTotalEveryMove() {
        return isCountTotalEveryMove;
    }

    public boolean getIsCountPlaceEveryMove() {
        return isCountPlaceEveryMove;
    }
}

File: app/src/main/java/com/example/dices2/Game.kt
package com.example.dices2

import android.content.SharedPreferences
import java.util.*

class Game private constructor() {
    private lateinit var mainActivity: MainActivity
    private lateinit var listActivity: ListActivity
    private var players: MutableList<Player> = ArrayList()
    private lateinit var currentCell: Cell
    var isCountTotalEveryMove = false
    var isCountPlaceEveryMove = false
    var isKeelScreenOn = true
    private var totalNumberOfMoves = 0
    // TODO добавить другие виды правил игры
    private val totalNumberOfRows = 16
    private var idCurrentPlayer = 0
    var isGameStarted = false
    private var currentPlayer: Player? = null

    init {
        Consts.savedCells = HashSet()
    }

    private val isGameOver: Boolean
        get() = totalNumberOfMoves <= 0

    // Ход игры
    private fun startGame() {
        isGameStarted = true
        getTotalNumberOfMoves()
        Consts.savedCells.clear()
        clearPlayersValues()
        idCurrentPlayer = 0
        currentPlayer = players[idCurrentPlayer]
        mainActivity.fillMainTable(players = players, player = currentPlayer)
        highlightNextPlayer()
    }

    private fun clearPlayersValues() {
        players.forEach {
            it.clearAllValues()
            it.clearPlace()
        }
    }

    private fun highlightNextPlayer() {
        val playersAmount = players.size
        idCurrentPlayer = if (idCurrentPlayer > playersAmount - 1) 0 else idCurrentPlayer
        val idHighlightOff = if (idCurrentPlayer == 0) playersAmount - 1 else idCurrentPlayer - 1
        val idHighlightOn = if (idCurrentPlayer > playersAmount - 1) 0 else idCurrentPlayer
        mainActivity.switchOffPlayersMove(player = players[idHighlightOff])
        currentPlayer = players[idHighlightOn]
        mainActivity.highlightPlayersMove(player = players[idHighlightOn])
        idCurrentPlayer++
    }

    private fun makeMove() {
        totalNumberOfMoves--
        highlightNextPlayer()
        mainActivity.setTextViewHighlight(currentCell = currentCell, isHighlight = true)
    }

    private fun makeEdit(cell: Cell) {
        if (currentCell.isEmpty) {
            mainActivity.setTextViewHighlight(currentCell = currentCell, isHighlight = false)
        } else {
            mainActivity.setTextViewHighlight(currentCell = currentCell, isHighlight = true)
        }
        if (cell.owner == currentPlayer) {
            mainActivity.highlightPlayersMove(player = currentPlayer)
        }
    }

    fun okButtonOnKeyboardActivityClicked(value: String, isEdit: Boolean) {
        val player = currentCell.owner
        val rowName = currentCell.row
        player.setValue(rowName = rowName, value = value)
        currentCell.value = value
        refreshCellsList()
        refreshFullSquarePokerTextViews(player = player)
        if (isEdit) {
            makeEdit(cell = currentCell)
        } else {
            makeMove()
        }
        showSchool(player = player)
        if (isCountTotalEveryMove) {
            showTotal(player = player)
        }
        if (isCountPlaceEveryMove) {
            showAllPlaces()
        }
        if (isGameOver) {
            showAllTotals()
            //isGameStarted = false
            showAllPlaces()
        }
    }

    private fun refreshFullSquarePokerTextViews(player: Player) {
        if (player.isThreeClassesInSchoolAreFinished && !player.isSecondNonSchoolPartAdded) {
            mainActivity.enableFullSquarePokerTextViews(player = player)
        }
        if (!player.isThreeClassesInSchoolAreFinished && player.isSecondNonSchoolPartAdded) {
            mainActivity.disableFullSquarePokerTextViews(player = player)
        }
    }

    fun setMainActivity(mainActivity: MainActivity) {
        this.mainActivity = mainActivity
        mainActivity.fillMainTable(players = players, player = currentPlayer)
    }

    fun setCurrentCell(cell: Cell) {
        currentCell = cell
    }

    private fun showSchool(player: Player) {
        mainActivity.showSchool(player = player)
    }

    private fun showTotal(player: Player) {
        mainActivity.showTotal(player = player)
    }

    private fun showAllTotals() {
        for (player in players) {
            showTotal(player = player)
        }
    }

    private fun showPlayerPlace(player: Player) {
        mainActivity.showPlayerPlace(player = player)
    }

    private fun showAllPlaces() {
        calcPlaces()
        for (player in players) {
            showPlayerPlace(player = player)
        }
    }

    // Работа с ListActivity
    fun setListActivity(listActivity: ListActivity) {
        this.listActivity = listActivity
        fillListView()
    }

    private fun fillListView() {
        listActivity.fillListView(players = players)
    }

    fun addPlayer(nameOfPlayer: String) {
        players.add(Player(name = nameOfPlayer))
        listActivity.refreshListView(players = players)
    }

    fun removePlayer(position: Int) {
        players.removeAt(position)
        listActivity.refreshListView(players = players)
    }

    fun clearListOfPlayersAndRefreshListView() {
        clearListOfPlayers()
        listActivity.refreshListView(players = players)
    }

    private fun clearListOfPlayers() {
        players.clear()
    }

    fun getNameOfPlayerByPosition(position: Int): String {
        return players[position].name
    }

    fun startGameButtonClicked() {
        startGame()
    }

    // Прочее

    private fun getTotalNumberOfMoves() {
        totalNumberOfMoves = players.size * totalNumberOfRows
    }

    fun isDuplicatePlayer(name: String): Boolean {
        for (player in players) {
            if (player.name == name) {
                return true
            }
        }
        return false
    }

    private fun calcPlaces() {
        val playersSortedByTotalValue: List<Player> = ArrayList(players)
        Collections.sort(playersSortedByTotalValue, Collections.reverseOrder())
        val currentPlayer = playersSortedByTotalValue[0]
        var place = 1
        var currentValue = currentPlayer.total
        currentPlayer.curPlace = place

        for (i in 1 until playersSortedByTotalValue.size) {
            val nextPlayer = playersSortedByTotalValue[i]
            if (currentValue == nextPlayer.total) {
                nextPlayer.curPlace = place
            } else {
                place++
                currentValue = nextPlayer.total
                nextPlayer.curPlace = place
            }
        }
    }

    private fun isEditInSchool(cell: Cell): Boolean {
        for (rowName in Consts.NAMES_OF_SCHOOL_ROWS) {
            if (cell.row == rowName) {
                return true
            }
        }
        return false
    }

    private fun refreshCellsList() {
        Consts.savedCells.add(currentCell)
    }

    fun getIsCountTotalEveryMove() = isCountTotalEveryMove

    fun getIsCountPlaceEveryMove() = isCountPlaceEveryMove

    fun getIsKeepScreenOn() = isKeelScreenOn

    fun savePlayers(pref: SharedPreferences) {
        val editor = pref.edit()
        val playersSet = getPlayersSet()
        editor.putStringSet(Consts.APP_PREFERENCES_PLAYERS, playersSet)
        editor.apply()
    }

    private fun getPlayersSet(): MutableSet<String> {
        val set = HashSet<String>()
        players.forEach { set.add(it.name) }
        return set
    }

    fun fillStartPlayers(playersSet: Set<String>?) {
        playersSet?.let {
            players.clear()
            playersSet.forEach {
                players.add(Player(it))
            }
        }
        if (players.size > 0) {
            startGame()
        }
    }

    private object HOLDER {
        val INSTANCE = Game()
    }

    companion object {
        val instance: Game by lazy { HOLDER.INSTANCE }
    }
}